docker.sh

```diff
@@ -0,0 +1,113 @@
+DEV_CONTAINER_NAME="hyped_dev"
+
+# Adapted from: https://medium.com/@wujido20/handling-flags-in-bash-scripts-4b06b4d0ed04
+
+# Function to display script usage
+usage() {
+ echo "Usage: $0 <yarn_script> [OPTIONS]"
+ echo "Options:"
+ echo " -b,  --build             Build with container"
+ echo " -d,  --dev               Run development container"
+ echo " -h,  --help              Display this help message"
+ echo " -r,  --rebuild           Rebuild the docker image before running the container"
+ echo " -c,  --clean             Makes a clean build directory"
+ echo " -cc, --cross-compile     Cross-compile for Raspberry Pi"
+}
+
+# From Kshitij
+if ! [ -x "$(command -v docker)" ]; then
+  echo '[!] Error: docker is not installed.' >&2
+  exit 1
+fi
+
+if ! [ -x "$(command -v docker-compose)" ]; then
+  echo '[!] Error: docker-compose is not installed.' >&2
+  exit 1
+fi
+
+# Check if image is built
+image=$( docker images -q $IMAGE_NAME 2> /dev/null )
+if [[ -z ${image} ]]; then
+  echo "[!] Building image"
+  docker build -t $IMAGE_NAME .
+else 
+  echo "[>] Image already built"
+fi
+
+# Default values for options
+rebuild=false
+clean=false
+cross_compile=false
+docker_build=false
+docker_dev=false
+
+# Function to handle options and arguments
+handle_options() {
+  while [ $# -gt 0 ]; do
+    case $1 in
+      -b | --build)
+        docker_build=true
+        ;;
+      -d | --dev)
+        docker_dev=true
+        ;;
+      -h | --help)
+        usage
+        exit 0
+        ;;
+      -r | --rebuild)
+        rebuild=true
+        ;;
+      -c | --clean)
+        clean=true
+        ;;
+      -cc | --cross-compile)
+        cross_compile=true
+        ;;
+      *)
+        echo "Invalid option: $1" >&2
+        usage
+        exit 1
+        ;;
+    esac
+    shift
+  done
+}
+
+# Main script execution
+handle_options "$@"
+
+if [ "$rebuild" = true ]; then
+  echo "Rebuild"
+  docker build -t $IMAGE_NAME docker
+fi
+
+if [ "$docker_build" = true ]; then
+  # Check if the container name already exists
+  build_container=$( docker ps -a -q --filter name=$BUILD_CONTAINER_NAME 2> /dev/null )
+
+  # If the container exists, remove it
+  if [[ -n ${build_container} ]]; then
+    echo "[!] Found existing container. Removing container"
+    docker rm $BUILD_CONTAINER_NAME
+  fi
+
+  docker run -it -e CLEAN=$clean -e CROSS_COMPILE=$cross_compile -e DIR=/home/$IMAGE_NAME --name $BUILD_CONTAINER_NAME -v $(pwd):/home/$IMAGE_NAME $IMAGE_NAME bash
+fi
+
+if [ "$docker_dev" = true ]; then
+  # Check if the container name already exists
+  dev_container=$( docker ps -a -q --filter name=$DEV_CONTAINER_NAME 2> /dev/null )
+
+  if [[ -n ${dev_container} ]]; then
+    echo "[>] Found existing container"
+    docker start -i $DEV_CONTAINER_NAME
+  else
+    echo "[!] No existing container found. Creating new container"
+    docker run -it --mount type=bind,source=$(pwd),target=/home/hyped/ --name $DEV_CONTAINER_NAME --entrypoint /bin/bash $IMAGE_NAME
```

Comment:
should I add the cross-compile and clean stuff to the dev environment? and are we happy with the approach of starting an existing container if it exists rather than removing old containers and creating a new container each time (as is done for the build)?

Reply:
I don't think we need cross compile in the dev env? I'm not sure. What is the practical difference between creating a new container or reusing an old one?

Reply:
i think for the env it is fine to reuse, idk what difference it makes for the build

Reply:
is it just that it doesn't preserve internal state between running? in that case we want a new container each time for builds, but the dev one idk. If it's mounting all the files from disk then there isn't really state and it's fine to reuse, but otherwise I'd be nervous about it drifting for some reason (primarily if you run `apt install` from within the container I guess)